src/Microsoft.AspNet.Mvc.Core/HttpResponseStreamWriter.cs

```diff
@@ -0,0 +1,280 @@
+                }
+
+                int remaining = _charBufferSize - _charBufferCount;
+                if (remaining > count)
+                {
+                    remaining = count;
+                }
+
+                value.CopyTo(
+                    sourceIndex: index,
+                    destination: _charBuffer,
+                    destinationIndex: _charBufferCount,
+                    count: remaining);
+
+                _charBufferCount += remaining;
+                index += remaining;
+                count -= remaining;
+            }
+        }
+
+        public override async Task WriteAsync(char value)
+        {
+            if (_charBufferCount == _charBufferSize)
+            {
+                await FlushAsync();
+            }
+
+            _charBuffer[_charBufferCount] = value;
+            _charBufferCount++;
+        }
+
+        public override async Task WriteAsync(char[] buffer, int index, int count)
+        {
+            if (buffer == null)
+            {
+                return;
+            }
+
+            while (count > 0)
+            {
+                if (_charBufferCount == _charBufferSize)
+                {
+                    await FlushAsync();
+                }
+
+                int remaining = _charBufferSize - _charBufferCount;
+                if (remaining > count)
+                {
+                    remaining = count;
+                }
+
+                Buffer.BlockCopy(
+                    src: buffer,
+                    srcOffset: index * sizeof(char),
+                    dst: _charBuffer,
+                    dstOffset: _charBufferCount * sizeof(char),
+                    count: remaining * sizeof(char));
+
+                _charBufferCount += remaining;
+                index += remaining;
+                count -= remaining;
+            }
+        }
+
+        public override async Task WriteAsync(string value)
+        {
+            if (value == null)
+            {
+                return;
+            }
+
+            int count = value.Length;
+            int index = 0;
+            while (count > 0)
+            {
+                if (_charBufferCount == _charBufferSize)
+                {
+                    await FlushAsync();
+                }
+
+                int charBufferFreeSpace = _charBufferSize - _charBufferCount;
+                if (charBufferFreeSpace > count)
+                {
+                    charBufferFreeSpace = count;
+                }
+
+                value.CopyTo(
+                    sourceIndex: index,
+                    destination: _charBuffer,
+                    destinationIndex: _charBufferCount,
+                    count: charBufferFreeSpace);
+
+                _charBufferCount += charBufferFreeSpace;
+                index += charBufferFreeSpace;
+                count -= charBufferFreeSpace;
+            }
+        }
+
+        // Do not flush the stream on Close/Dispose, as this will cause response to be
```

Comment:
Oh nevermind. This makes sense. Maybe put this closer to the `Close` \ `Dispose methods.